### PR TITLE
[close #33907] Error when using "recyclable" cache keys with a store that does not support it

### LINKED ISSUE #33907: The cache_version is not used by dalli when reading and writing keys of objects (memcache client)
### Update

The `dalli_store` does not work with rails anymore and you should use `mem_cache_store` instead. Bellow is my reproduction and report that I wrote before getting this information.

### Steps to reproduce

```bash
git clone https://github.com/schneems/dalli-cache-bug
cd dalli-cache-bug
bundle install
rails console
```

then run this code:

```ruby
ModelWithKeyAndVersion = Struct.new(:cache_key, :cache_version)
model_1_cache_1 = ModelWithKeyAndVersion.new("model/1", "1")
model_1_cache_2 = ModelWithKeyAndVersion.new("model/1", "2")

Rails.cache.fetch(model_1_cache_1) { "woohoo" }
puts Rails.cache.read(model_1_cache_1) # => "woohoo"
puts Rails.cache.read(model_1_cache_2) # => "woohoo"
```

### Expected behavior
 
Different cache versions should produce different cache keys and therefore different cache results.

### Actual behavior

Instead, dalli does not use the `cache_version` at all so you get the same result even if the `cache_version` changes.

### System configuration

**Rails version**: 5.2.1 && master

**Ruby version**: 2.5.1

## Notes

Other stores implemented directly in Rails such as MemoryStore utilize ActiveSupport::Cache which does manually expand `cache_version` https://github.com/rails/rails/blob/11581afc01927565d57dd0e0156e983f92f8c2e0/activesupport/lib/active_support/cache.rb#L661. 

Right now if you're using dalli and rails then rev-ing the cache_version of an object does not invalidate the cache. 😵🙊🙈🙉😵

## More reproductions

I added a view to my example app to show that this issue does affect view caching and not just `Rails.cache` calls. Details are in the readme. Also, here's a video of me reproducing the behavior in production on [CodeTriage](https://www.codetriage.com).

screenflow

### PR DESCRIPTION
**Pull request by schneems**
If you are using the "in cache versioning" also known as "recyclable cache keys" the cache store must be aware of this scheme, otherwise you will generate cache entries that never invalidate.

This PR adds a check to the initialization process to ensure that if recyclable cache keys are being used via 

```
config.active_record.cache_versioning = true
```

Then the cache store needs to show that it supports this versioning scheme. Cache stores can let Rails know that they support this scheme by adding a method `supports_in_cache_versioning?` and returning true.

### CODE REVIEW

**Review by rafaelfranca — APPROVED** [review]

File: activerecord/lib/active_record/railtie.rb [hunk 1]
```diff
@@ -88,6 +88,29 @@ class Railtie < Rails::Railtie # :nodoc:
       end
     end
 
+    initializer "Check for cache versioning support" do
+      config.after_initialize do |app|
+        ActiveSupport.on_load(:active_record) do
+          if app.config.active_record.cache_versioning && defined?(Rails) && Rails.cache
```
[the remark this answers is not in the document]
Comment (schneems) [reply activerecord/lib/active_record/railtie.rb]: good point, it's in a..railtie. Fixed.

File: activerecord/lib/active_record/railtie.rb [hunk 2]
```diff
@@ -88,6 +88,29 @@ class Railtie < Rails::Railtie # :nodoc:
...
+    initializer "Check for cache versioning support" do
+      config.after_initialize do |app|
+        ActiveSupport.on_load(:active_record) do
+          if app.config.active_record.cache_versioning && Rails.cache
+            unless Rails.cache.try(:supports_in_cache_versioning?)
+              raise <<-end_error
+
+You're using a cache store `#{Rails.cache.class}` that does not support
+"recyclable" cache keys, also known as "in cache versioning". To
+fix this issue either disable "recyclable" cache keys by setting:
```
Comment (jeremy) [inline activerecord/lib/active_record/railtie.rb]: Let's stick with cache versioning terminology as a conceptual anchor and rely on the guides to discuss in detail. We can drop quoted terms. Let's be clear about what we're asking of the cache store, too, e.g.

> You're using a cache store that doesn't support native cache versioning. Your best option is to upgrade to a newer version of `#{Rails.cache.class}` that supports cache versioning (`#{Rails.cache.class}.supports_cache_versioning? #=> true`). Next best, switch to a different cache store that does support cache versioning: https://guides.rubyonrails.org/caching_with_rails.html#cache-stores. To keep using the current cache store, you can turn off cache versioning entirely: `config.active_record.cache_versioning = false`.

File: activerecord/lib/active_record/railtie.rb [hunk 3]
```diff
@@ -88,6 +88,29 @@ class Railtie < Rails::Railtie # :nodoc:
       end
     end
 
+    initializer "Check for cache versioning support" do
+      config.after_initialize do |app|
+        ActiveSupport.on_load(:active_record) do
+          if app.config.active_record.cache_versioning && Rails.cache
+            unless Rails.cache.try(:supports_in_cache_versioning?)
```
Comment (jeremy) [inline activerecord/lib/active_record/railtie.rb]: Let's use a class method and normalize naming: `Rails.cache.class.supports_cache_versioning?`

File: activesupport/lib/active_support/cache/file_store.rb [hunk 4]
```diff
@@ -26,6 +26,13 @@ def initialize(cache_path, options = nil)
         @cache_path = cache_path.to_s
       end
 
+      # Advertise that this cache store can be used
+      # with "recyclable cache keys" otherwise known
+      # as cache versioning.
+      def supports_in_cache_versioning?
+        true
+      end
+
```
Comment (jeremy) [inline activesupport/lib/active_support/cache/file_store.rb]: Let's do a `supports_cache_versioning?` class method for these.

Comment may be unwrapped to `# Advertise cache versioning support.`

(Worth noting that this duck typing won't cover subclasses of these cache stores.)

File: railties/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,8 @@
+*   Raise an error when "recyclable cache keys" are being used by a cache store
+    that does not explicitly support it.
```
Comment (jeremy) [inline railties/CHANGELOG.md]: Let's describe what cache stores need to do (implement `supports_cache_versioning` class method) to advertise support.

Comment (schneems) [reply railties/CHANGELOG.md]: For this what are you thinking? Maybe a section in the guide?

File: activerecord/lib/active_record/railtie.rb (hunk 3 above)
Comment (schneems) [reply activerecord/lib/active_record/railtie.rb]: I made a comment about the naming in basecamp. I think it's confusing. All caches are versioned already via cache key, the conceptual change is that the versioning is now __in__ the cache which is what i'm trying to convey here with that word. I can standardize on this though.

**Pull request merged**